### PR TITLE
docs: comprehensive batch and gang scheduling documentation

### PR DESCRIPTION
**Pull request by wcarrollrai**
## Description

This PR provides comprehensive documentation for batch and gang scheduling in KAI-Scheduler, expanding the minimal existing documentation (24 lines) to cover all 9 supported workload types with practical, production-ready examples (640 lines).

### What Changed

**Documentation Enhancements:**
- Expanded `docs/batch/README.md` with clear definitions of batch vs gang scheduling
- Added 9 workload type examples in new `docs/batch/examples/` directory
- Included external requirements with operator install commands
- Added cross-links to topology-aware scheduling and pod-grouper documentation

**New Example Files (9 workload types):**
1. `job.yaml` - Standard Kubernetes Job (batch scheduling)
2. `pytorchjob.yaml` - PyTorch distributed training with MNIST
3. `mpijob.yaml` - MPI distributed computing (pi calculation)
4. `tfjob.yaml` - TensorFlow distributed training with MNIST
5. `xgboostjob.yaml` - XGBoost distributed training on Iris dataset
6. `jaxjob.yaml` - JAX distributed training on MNIST
7. `rayjob.yaml` - Ray distributed computing cluster
8. `jobset.yaml` - Multi-job workflow coordination
9. `sparkapplication.yaml` - Apache Spark data processing

**Key Features:**
- All examples use `schedulerName: kai-scheduler` with proper queue labels
- Only non-gcr.io container images (all validated and pullable)
- Practical, runnable examples (not placeholder sleep commands)
- Copyright headers on all files
- kubectl validation performed on all manifests

**Cleanup:**
- Removed legacy `batch-job.yaml` (replaced by examples/job.yaml)
- Moved and enhanced `pytorch-job.yaml` to examples/pytorchjob.yaml

### Why This Change

The existing batch documentation was minimal and only covered 2 workload types with placeholder examples. This made it difficult for users to:
- Understand the difference between batch and gang scheduling
- Know which workload types are supported
- Get started quickly with practical examples
- Understand external requirements for third-party CRDs

This PR addresses these gaps by providing comprehensive, copy/paste-ready documentation.

## Related Issues

N/A - Documentation enhancement

## Checklist

- [x] Self-reviewed
- [x] Added/updated tests (if needed) - N/A for documentation
- [x] Updated documentation (if needed) - This PR IS documentation

## Breaking Changes

None. This is a documentation-only change that enhances existing content without modifying any code or APIs.

### Validation Performed

**kubectl dry-run validation:**
- ✅ Native K8s resources (Job) pass validation
- ✅ Third-party CRDs show expected "CRD not installed" errors (validates YAML syntax)

**Image validation:**
All container images verified as non-gcr.io, publicly accessible, and manifest-inspectable:
- `ubuntu:latest` (Docker Hub)
- `ghcr.io/kubeflow/training-v1/pytorch-dist-mnist:latest` (GitHub Container Registry)
- `mpioperator/mpi-pi:openmpi` (Docker Hub)
- `kubeflow/tf-mnist-with-summaries:latest` (Docker Hub)
- `docker.io/kubeflow/xgboost-dist-iris:latest` (Docker Hub)
- `docker.io/kubeflow/jax-mnist:latest` (Docker Hub)
- `rayproject/ray:2.46.0` (Docker Hub)
- `busybox:1.36` (Docker Hub)
- `apache/spark:3.5.0` (Docker Hub)

### Reviewer Guidance

**What to Review:**
- README structure and clarity
- Example completeness (all 9 supported workload types covered)
- Accuracy of operator install commands
- Appropriateness of selected container images

**Related Documentation:**
- [Pod Grouper Technical Details](docs/developer/pod-grouper.md) - Linked from batch README
- [Topology-Aware Scheduling](docs/topology/README.md) - Linked for distributed workloads

### Before & After

**Before:** 24 lines, 2 workload types, sleep-based placeholders
**After:** 640 lines, 9 workload types, practical runnable examples

The new documentation provides:
- Clear batch vs gang scheduling definitions
- Comprehensive coverage with inline YAML + raw files
- External requirements with install commands
- Cross-links to related documentation

## Summary by CodeRabbit

* **Documentation**
  * Expanded batch scheduling guidance with independent and gang scheduling, workload configuration, topology awareness, PodGroups, supported workload types, and practical examples.
  * Updated Ray configuration guidance with links to standalone manifests and simplified topology instructions.
* **Examples**
  * Added ready-to-use examples for Kubernetes Jobs, JobSets, MPI, JAX, TensorFlow, XGBoost, Spark, and Ray workloads.
  * Updated PyTorch and RayJob examples with clearer metadata, resource settings, and embedded sample code.
  * Removed the outdated standalone batch Job example.

### CODE REVIEW

**Comment by enoodle** [comment]
@wcarrollrai Thanks,
Maybe it will be more easy to read through if you link in the README to the example files instead of having the whole example yamls there?

**Comment by wcarrollrai** [comment]
## Updates Based on Reviewer Feedback

  Thanks for the feedback! I've refactored the README to improve readability and reduce maintenance burden.

  ### Changes Made

  **Improved Readability:**
  - Reduced README from 640 lines to 198 lines (69% reduction)
  - Removed all inline YAML examples in favor of links to example files
  - Added a quick reference table at the top for easy navigation
  - Reformatted each workload section with bullet lists for better scannability

  **Reduced Maintenance:**
  - Replaced hardcoded installation commands with links to official documentation
  - This ensures users always get the latest installation instructions from upstream
  - No need to update this repo when operators change their installation procedures

  ### What Stayed the Same

  - All 9 example YAML files remain unchanged in `docs/batch/examples/`
  - All descriptions, scheduling behaviors, and external requirements preserved
  - Topology-aware scheduling section maintained

  ### Before & After

  **Before:**
  - 640-line README with 9 inline YAML blocks
  - Hardcoded `kubectl apply` and `helm install` commands

  **After:**
  - 198-line README with links to examples
  - Quick reference table
  - Bullet list format for each workload type
  - Links to official installation guides

  The documentation is now more maintainable and easier to scan while preserving all essential information.

**Comment by enoodle** [comment]
@wcarrollrai Can this also somehow absorb this example with its extra documentation: https://github.com/NVIDIA/KAI-Scheduler/tree/main/examples/ray

File: docs/batch/README.md [hunk 1]
```diff
@@ -1,23 +1,163 @@
...
 
+KAI Scheduler provides sophisticated workload scheduling with support for both batch scheduling and gang scheduling. The scheduler automatically detects the workload type and applies the appropriate scheduling strategy through the PodGrouper component, which creates PodGroup custom resources to coordinate pod scheduling.
 
-## PyTorchJob
-To run in a distributed way across multiple pods, you can use PyTorchJob.
+## Definitions
 
-### Prerequisites
-This requires the [kubeflow-training-operator-v1](https://www.kubeflow.org/docs/components/trainer/legacy-v1/) to be installed in the cluster.
+### Batch Scheduling
```
Comment (gshaibi) [inline docs/batch/README.md]: I'm not familiar with this term in the context of independent scheduling. 
What do you think about renaming it to "independent scheduling"/"pod-by-pod scheduling" or something alike?

**Review by gshaibi — COMMENTED** [review]
@wcarrollrai thanks for contributing!!!

**Comment by wcarrollrai** [comment]
@enoodle - Sure. Let me see what we can do to best handle this.

File: docs/batch/README.md (hunk 1 above)
Comment (wcarrollrai) [reply docs/batch/README.md]: I see your point. I'll get this updated.

**Comment by gshaibi** [comment]
Hey @wcarrollrai , we already have an `examples/batch` directory under the root of the repository which I think might be a better place for these manifests. What do you think?

**Comment by wcarrollrai** [comment]
Hi @gshaibi - Happy to move these examples to the `examples/batch` under root. I'll update locations and paths in the README. Also will take some time this week to ensure that each and every example has been tested.

**Comment by wcarrollrai** [comment]
@gshaibi - Examples moved and all tested. They all pass in their current state assuming all pre-reqs are installed in the cluster.

File: docs/batch/README.md [hunk 2]
```diff
@@ -47,12 +42,197 @@ Behavior notes:
 ## PyTorchJob
 To run in a distributed way across multiple pods, you can use PyTorchJob.
```
Comment (gshaibi) [inline docs/batch/README.md]: I think this section should now be deleted, what do you think?

File: docs/batch/README.md [hunk 3]
```diff
@@ -47,12 +42,197 @@ Behavior notes:
...
+| [SparkApplication](#sparkapplication-spark-operator) | Gang | Spark Operator | [sparkapplication.yaml](../../examples/batch/sparkapplication.yaml) |
+
+### Standard Kubernetes Job
+
+Standard Kubernetes Jobs run batch workloads where pods can be scheduled independently.
+
+- **Scheduling Behavior:** Independent scheduling (pods scheduled independently)
+- **External Requirements:** None (native Kubernetes resource)
+- **Example:** [examples/batch/job.yaml](../../examples/batch/job.yaml)
+- **Apply:** `kubectl apply -f examples/batch/job.yaml`
```
Comment (gshaibi) [inline docs/batch/README.md]: nit - I don't think these `kubectl apply...` directions help much, consider removing them if you feel the same.

File: docs/batch/README.md [hunk 4]
```diff
@@ -1,13 +1,8 @@
...
-To run a simple batch job with multiple pods that will be scheduled separately, run the following command:
-```
-kubectl apply -f batch-job.yaml
-```
-This will create 2 pods that will be scheduled separately. Both pods will either run at the same time or sequentially, depending on the available resources in the cluster.
+## Overview
 
+KAI Scheduler provides sophisticated workload scheduling with support for both independent scheduling and gang scheduling. The scheduler automatically detects the workload type and applies the appropriate scheduling strategy through the PodGrouper component, which creates PodGroup custom resources to coordinate pod scheduling.
 
 ## Min Member Override
```
Comment (gshaibi) [inline docs/batch/README.md]: This is an advanced topic and should be after the `Definitions` and `How it works` sections in my opinion.

File: docs/batch/README.md
Comment (gshaibi) [inline docs/batch/README.md]: This is an advanced topic and should be after the `Definitions` and `How it works` sections in my opinion.

File: examples/ray/README.md [hunk 5]
```diff
@@ -19,179 +19,23 @@ helm install kuberay-operator kuberay/kuberay-operator \
...
 |-------|----------|-------|-------------|
 | `kai.scheduler/queue` | `metadata.labels` (on RayJob/RayCluster) | Queue name (e.g., `default-queue`) | Assigns workload to a KAI queue |
-| `schedulerName` | `spec.template.spec` (on each pod template) | `kai-scheduler` | Routes pods to KAI scheduler |
-
-## Topology-Aware Scheduling (Optional)
-
-To enable topology-aware scheduling for Ray subgroups, add topology annotations on the pod templates:
-
-- `spec.headGroupSpec.template.metadata.annotations`
-- `spec.workerGroupSpecs[*].template.metadata.annotations`
```
Comment (gshaibi) [inline examples/ray/README.md:135]: Where is this information documented now?

File: examples/ray/README.md [hunk 6]
```diff
@@ -19,179 +19,23 @@ helm install kuberay-operator kuberay/kuberay-operator \
...
-    kai.scheduler/queue: "default-queue"
-spec:
-  rayVersion: '2.54.0'
-  headGroupSpec:
-    template:
-      metadata:
-        annotations:
-          kai.scheduler/topology: "cluster-topology"
-          kai.scheduler/topology-required-placement: "rack"
-          kai.scheduler/topology-preferred-placement: "node"
```
Comment (gshaibi) [inline examples/ray/README.md:160]: In my opinion, this has value here.

File: examples/ray/rayjob.yaml [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+          containers:
+          - name: ray-head
+            image: rayproject/ray:2.46.0
+            volumeMounts:
+            - name: rayjob-code
+              mountPath: /home/ray/samples
+            resources:
+              limits:
+                cpu: "1"
+                nvidia.com/gpu: "1"
```
Comment (gshaibi) [inline examples/ray/rayjob.yaml]: Please comment that this is optional, like it was on the file you removed.

File: examples/batch/jaxjob.yaml [hunk 8]
```diff
@@ -0,0 +1,32 @@
+# Copyright 2025 NVIDIA CORPORATION
```
Comment (gshaibi) [inline examples/batch/jaxjob.yaml]: Please update copyright year

**Review by gshaibi — COMMENTED** [review]
Thanks @wcarrollrai for the verifications!
Left a couple more small comments.

File: docs/batch/README.md (hunk 2 above)
Comment (wcarrollrai) [reply docs/batch/README.md]: Agreed. Will remove.

File: docs/batch/README.md (hunk 3 above)
Comment (wcarrollrai) [reply docs/batch/README.md]: Agreed. Will clean all of these up.

File: examples/batch/jaxjob.yaml (hunk 8 above)
Comment (wcarrollrai) [reply examples/batch/jaxjob.yaml]: Updated.

File: examples/ray/rayjob.yaml (hunk 7 above)
Comment (wcarrollrai) [reply examples/ray/rayjob.yaml]: Comments restored in both rayjob and raycluster manifests

File: examples/ray/README.md (hunk 5 above)
Comment (wcarrollrai) [reply examples/ray/README.md:135]: Good point. It was not recreated anywhere. I'll readd this section

File: examples/ray/README.md (hunk 6 above)
Comment (wcarrollrai) [reply examples/ray/README.md:160]: Agreed. Will update with this information readded.

File: docs/batch/README.md (hunk 4 above)
Comment (wcarrollrai) [reply docs/batch/README.md]: Moved to be after the `Definitions` and `How it works` sections.

File: docs/batch/README.md
Comment (wcarrollrai) [reply docs/batch/README.md]: Moved to be after the `Definitions` and `How it works` sections.

**Comment by wcarrollrai** [comment]
@gshaibi - Thanks for the review and comments. I've addressed each of the comments and made updates. Please take a look and let me know if you have any additional thoughts.

**Comment by gshaibi** [comment]
@coderabbitai review

File: docs/batch/README.md [hunk 9]
```diff
@@ -1,24 +1,56 @@
...
+|-------|----------|-------|-------------|
+| `kai.scheduler/queue` | `metadata.labels` | Queue name (e.g., `default-queue`) | Assigns workload to a KAI queue |
+| `schedulerName` | Pod template spec | `kai-scheduler` | Routes pods to KAI scheduler |
 
+**Note:** For workloads with multiple pod templates (e.g., Ray head and workers, Spark driver and executors), you must set `schedulerName: kai-scheduler` in each pod template spec.
 
 ## Min Member Override
-To require a minimum number of pods to be scheduled together (gang scheduling) for a batch Job or JobSet, use the `kai.scheduler/batch-min-member` annotation on the Job or JobSet resource:
-```
-kubectl apply -f batch-job-min-member.yaml
```
Comment (gshaibi) [inline docs/batch/README.md:15]: Why did we remove this reference? Can we have a reference like you did for external-podgroup-job.yaml?
Also, please remove the docs/batch/batch-job-min-member.yaml file as it is not referenced anymore.

**Review by gshaibi — COMMENTED** [review]
Thanks @wcarrollrai , one more comment.

**Review by enoodle — APPROVED** [review]
